File: Data Structure/Vector/vector.py
# Data Structure
# Vector

class Vector:
    def __init__(self, vec):
        self.vec = vec
        self.size = len(vec)
        
    def __del__(self):
        del self.vec
        
    def getSize(self):
        print("Vector size: ", self.size)
        
    def insert(self, index, val):
        if index < 0 or index > self.size:
            return -1
        self.vec.insert(index, val)
        
    def push_back(self, val):
        self.vec.append(val)
        
    def pop_back(self):
        self.vec.pop()
        
    def erase(self, index):
        if index < 0 or index > self.size:
            return -1
        self.vec.pop(index)
        
    def sort(self):
        self.vec.sort()
        
    def reverse(self):
        self.vec.reverse()
        
    def print(self):
        print("Value inside vector: ")
        for i in self.vec:
            print(i, end = " ")
        print()
        
        
if __name__ == "__main__":
    vec = Vector([])
    
    vec.getSize()
    
    vec.print()
    
    for i in range(5):
        vec.push_back(i)
        
    vec.print()
    
    vec.insert(0, -2)
    
    vec.print()
    
    vec.pop_back()
    
    vec.print()
    
    vec.erase(2)
    
    vec.print()
    
    vec.insert(2, 16)
    
    vec.sort()
    
    vec.print()
    
    vec.reverse()
    
    vec.print()

File: Data Structure/Vector/vector.cpp
#include <iostream>
#include <vector>
#include <algorithm>

class Vector {
private:
    std::vector<int> vec;

public:
    Vector() : vec() {}

    void push_back(int val) {
        vec.push_back(val);
    }

    void pop_back() {
        vec.pop_back();
    }

    void insert(int index, int val) {
        if (index < 0 || index > vec.size()) 
            std::cout << "Index error \n";
        vec.insert(vec.begin() + index, val);
    }

    void remove(int index) {
        if (index < 0 || index > vec.size()) 
            std::cout << "Index error \n";
        vec.erase(vec.begin() + index);
    }

    void print() {
        std::cout << "Values inside vector: ";
        for (auto i : vec) 
            std::cout << i << " ";
        std::cout << "\n";
    }

    void sort() {
        std::sort(vec.begin(), vec.end());
    }

    void reverse() {
        std::reverse(vec.begin(), vec.end());
    }
    
    void getSize() {
        std::cout << "Vector size: " << vec.size() << "\n";
    }
};

int main(int argc, char **argv) {
    Vector vector;
    
    vector.getSize();

    vector.push_back(5);
    vector.push_back(3);
    vector.push_back(7);
    vector.push_back(1);
    
    vector.getSize();

    vector.print();

    vector.sort();

    vector.print();

    vector.reverse();

    vector.print();
    
    vector.insert(2, 10);
    
    vector.print();
    
    vector.pop_back();
    
    vector.print();
    
    vector.remove(0);

    vector.print();
    
    return 0;
}
